stable/database/files/backup_hooks.py

```diff
@@ -400,105 +434,262 @@ def get_query_params(query_str):
             raise UserError(
                 "Multiple values specified for query parameter '{}'".format(key)
             )
-        # Try to decode value as some token other than string, otherwise fall
-        # back to string
-        try:
-            as_map[key] = json.loads(values[0])
-        except json.decoder.JSONDecodeError:
-            as_map[key] = values[0]
+        as_map[key] = values[0]
     return as_map
 
 
-def dispatch_request(environ):
-    # Parse URL and make sure it contains at least one path component
-    uri = wsgiref.util.request_uri(environ)
-    parsed = urllib.parse.urlparse(uri)
-    path = parsed.path
-    if path.startswith("/"):
-        path = path[1:]
-    components = path.split("/")
-    if not components:
-        raise UserError("No path provided")
+def normalize_path(path):
+    return path[1:] if path.startswith("/") else path
+
+
+class ScriptHandler(object):
+
+    DEFAULT_STATUS_MAPPING = {
+        "0": 200,
+        "*": 400,
+    }
+
+    def __init__(self, **kwargs):
+        self.method = kwargs["method"]
+        self.path = kwargs["path"]
+        self.script = kwargs["script"]
+        self.status_mappings = dict(self.DEFAULT_STATUS_MAPPING)
+        if "statusMappings" in kwargs:
+            self.status_mappings.update(kwargs["statusMappings"])
+        self.path_params = {}
 
+    def matches(self, method, path_components):
+        # Check that method matches
+        if method != self.method:
+            return False
+
+        # Check that path matches and extract path parameters
+        handler_components = normalize_path(self.path).split("/")
+        for i in range(0, len(path_components)):
+            # If we have more path components in request than in handler path,
+            # then request does not match
+            if i >= len(handler_components):
+                return False
+            # If handler path component is a parameter, extract value from
+            # request path
+            handler_component = handler_components[i]
+            if handler_component.startswith("{") and handler_component.endswith("}"):
+                self.path_params[handler_component[1:-1]] = path_components[i]
```

Comment:
It looks like `ScriptHandler` is shared among different request threads (not sure if the service is multi-threaded).
Storing the path parameters might be problematic for concurrent requests.

Reply:
Good catch. The only mutable field is `path_params`, so I can just create a copy of the handler before invoking `match()`.